File: apps/ios/PassAlarm/Domain/Entity/Weekday.swift
import Foundation

struct WeekdaySet: OptionSet, Sendable {
    let rawValue: UInt8

    static let monday    = WeekdaySet(rawValue: 1 << 0)
    static let tuesday   = WeekdaySet(rawValue: 1 << 1)
    static let wednesday = WeekdaySet(rawValue: 1 << 2)
    static let thursday  = WeekdaySet(rawValue: 1 << 3)
    static let friday    = WeekdaySet(rawValue: 1 << 4)
    static let saturday  = WeekdaySet(rawValue: 1 << 5)
    static let sunday    = WeekdaySet(rawValue: 1 << 6)

    static let weekdays: WeekdaySet = [.monday, .tuesday, .wednesday, .thursday, .friday]

    static let allDays: [(WeekdaySet, String, Int)] = [
        (.monday, "月", 2),
        (.tuesday, "火", 3),
        (.wednesday, "水", 4),
        (.thursday, "木", 5),
        (.friday, "金", 6),
        (.saturday, "土", 7),
        (.sunday, "日", 1),
    ]

    func containsCalendarWeekday(_ calendarWeekday: Int) -> Bool {
        for (day, _, cw) in WeekdaySet.allDays {
            if cw == calendarWeekday && self.contains(day) {
                return true
            }
        }
        return false
    }
}

File: apps/android/app/src/main/kotlin/com/nogeass/passalarm/domain/model/Weekday.kt
package com.nogeass.passalarm.domain.model

enum class Weekday(val bit: Int, val label: String, val calendarValue: Int) {
    MONDAY(1, "月", java.util.Calendar.MONDAY),
    TUESDAY(2, "火", java.util.Calendar.TUESDAY),
    WEDNESDAY(4, "水", java.util.Calendar.WEDNESDAY),
    THURSDAY(8, "木", java.util.Calendar.THURSDAY),
    FRIDAY(16, "金", java.util.Calendar.FRIDAY),
    SATURDAY(32, "土", java.util.Calendar.SATURDAY),
    SUNDAY(64, "日", java.util.Calendar.SUNDAY);

    companion object {
        fun fromMask(mask: Int): Set<Weekday> =
            entries.filter { mask and it.bit != 0 }.toSet()

        fun toMask(days: Set<Weekday>): Int =
            days.fold(0) { acc, day -> acc or day.bit }

        fun containsCalendarDay(mask: Int, calendarDay: Int): Boolean =
            entries.any { it.calendarValue == calendarDay && mask and it.bit != 0 }
    }
}
